oci/pull.bzl

```diff
@@ -0,0 +1,385 @@
+    else:
+        index_mf = """\
+{
+   "schemaVersion": 2,
+   "manifests": [
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "size": %s,
+         "digest": "%s"
+      }
+   ]
+}""" % (image_mf_len, image_digest)
+
+    rctx.file("BUILD.bazel", content = _build_file.format(
+        name = rctx.attr.name,
+        tars = tars,
+        index_content = index_mf,
+        manifest_file = image_mf_file,
+        config_file = image_config_file,
+    ))
+
+oci_pull_rule = repository_rule(
+    implementation = _oci_pull_impl,
+    attrs = {
+        "image": attr.string(doc = "The name of the image we are fetching, e.g. gcr.io/distroless/static", mandatory = True),
+        "digest": attr.string(doc = "The digest of the manifest file", mandatory = True),
+        "platform": attr.string(doc = "platform in `os/arch` format, for multi-arch images"),
+    },
+)
+
+_alias_target = """\
+alias(
+    name = "{name}",
+    actual = select(
+        {platform_map}
+    ),
+    visibility = ["//visibility:public"],
+)
+"""
+
+def _oci_alias_impl(rctx):
+    rctx.file("BUILD.bazel", content = _alias_target.format(
+        name = rctx.attr.name,
+        platform_map = {str(k): v for k, v in rctx.attr.platforms.items()},
+    ))
+
+oci_alias = repository_rule(
+    implementation = _oci_alias_impl,
+    attrs = {
+        "platforms": attr.label_keyed_string_dict(),
+    },
+)
+
+_latest_build = """\
+load("@aspect_bazel_lib//lib:jq.bzl", "jq")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+
+jq(
+    name = "platforms",
+    srcs = ["manifest_list.json"],
+    filter = "[(.manifests // [])[] | .platform | .os + \\"/\\" + .architecture]",
+    # Print without newlines because it's too hard to indent that to fit under the generated
+    # starlark code below.
+    args = ["--compact-output"],
+)
+
+jq(
+    name = "mediaType",
+    srcs = ["manifest_list.json"],
+    filter = ".mediaType",
+    args = ["--raw-output"],
+)
+
+sh_binary(
+    name = "pin",
+    srcs = ["pin.sh"],
+    data = [
+        ":mediaType",
+        ":platforms",
+        "@bazel_tools//tools/bash/runfiles",
+    ],
+)
+"""
+
+_pin_sh = """\
+#!/usr/bin/env bash
+{rlocation}
+
+mediaType="$(cat $(rlocation {name}/mediaType.json))"
+echo -e "Replace your '{name}' declaration with the following:\n"
+
+cat <<EOF
+oci_pull(
+    name = "{name}",
+    digest = "sha256:{digest}",
+    image = "{image}",
+EOF
+
+[[ $mediaType == "{manifestListType}" ]] && cat <<EOF
+    # Listing of all platforms that were found in the image manifest.
+    # You may remove any that you don't use.
+    platforms = $(cat $(rlocation {name}/platforms.json)),
+EOF
+
+echo ")"
+"""
+
+def _pull_latest_impl(rctx):
+    """Download the 'latest' tag and create a repository that can produce pinning instructions"""
+    _download(rctx, "latest", "manifest_list.json", "manifests")
+    result = rctx.execute(["shasum", "-a", "256", "manifest_list.json"])
```

Comment:
nit: we should use the one from `bazel-lib/coreutils` that we recently introduced.

Reply:
right... won't work since we don't have aliases for coreutils. nvm